core/tee/tee_ree_fs.c

```diff
@@ -497,9 +498,19 @@ static TEE_Result open_dirh(struct tee_fs_dirfile_dirh **dirh)
 		return res;
 
 	res = tee_fs_dirfile_open(false, hashp, &ree_dirf_ops, dirh);
-	if (res == TEE_ERROR_ITEM_NOT_FOUND)
+	if (res == TEE_ERROR_ITEM_NOT_FOUND) {
 		res = tee_fs_dirfile_open(true, NULL, &ree_dirf_ops, dirh);
 
+		if (IS_ENABLED(CFG_REE_FS_ALLOW_RESET) && !res && hashp) {
+			DMSG("Clear REE FS hash in RPMB");
+			res = rpmb_fs_ops.truncate(ree_fs_rpmb_fh, 0);
```

Comment:
We have a race here. But perhaps we don't care since it's testing code only?

Reply:
i don't wan't to introduce bugs in the core. Sorry I don't see the reace.

Would you prefer we wipe the hash data in RPMB before we create the file in REE FS? Like:
```c
	res = tee_fs_dirfile_open(false, hashp, &ree_dirf_ops, dirh);
	if (res == TEE_ERROR_ITEM_NOT_FOUND) {
		if (IS_ENABLED(CFG_REE_FS_ALLOW_RESET) && hashp) {
			DMSG("Clear REE FS hash in RPMB");
			res = rpmb_fs_ops.truncate(ree_fs_rpmb_fh, 0);
			if (res) {
				DMSG("Clear REE FS hash failed: %#"PRIx32, res);
				res = TEE_ERROR_SECURITY;
				goto out;
			}
		}

		res = tee_fs_dirfile_open(true, NULL, &ree_dirf_ops, dirh);
	}
```

Reply:
Yes something like that.
The race I was after is if execution is stopped after the dirfile is created and we restart we wouldn't be able to reach this code path again. It could of course be fixed by the user removing the dirfile again, but I'd prefer doing it right as it's easier to understand the code then.

Reply:
Ok, agree. Let's clear the hash first.
Note that removing REE FS content here will work only if OP-TEE has not yet accessed it since it booted up.

Reply:
By the way, I have another change I was to propose, i think it simplifies the sequence. 
When CFG_REE_FS_ALLOW_RESET=n (default config),  there is a REE FS hash in the RPMB and dirf.db is not found, then we should report the error straight, not attempt to create back dirfs.db file and report the hash mismatch later.

I'll append the commit in this P-R.

Reply:
That's going to result in a chicken-egg problem, right?

Reply:
> That's going to result in a chicken-egg problem, right?

Talking about commit "core: REE FS: report earlier unexpected REE FS reset"?
The idea is to report the mismatch as early as possible. When looking into this part, it looks strange OP-TEE requests the creation of a file and later fails on its content.

Reply:
Sorry, I misunderstood what you meant.